compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/libgraal/LibGraalFeature.java

```diff
@@ -0,0 +1,420 @@
+ */
+package jdk.graal.compiler.libgraal;
+
+import java.lang.module.ModuleDescriptor;
+import java.lang.reflect.Field;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+
+import jdk.graal.compiler.core.common.spi.ForeignCallSignature;
+import jdk.graal.compiler.graph.NodeClass;
+import jdk.graal.compiler.hotspot.EncodedSnippets;
+import jdk.graal.compiler.hotspot.HotSpotForeignCallLinkage;
+import jdk.graal.compiler.hotspot.HotSpotReplacementsImpl;
+import jdk.graal.compiler.libgraal.truffle.LibGraalTruffleHostEnvironmentLookup;
+import jdk.graal.compiler.options.OptionsParser;
+import jdk.graal.compiler.serviceprovider.GraalServices;
+import jdk.graal.compiler.truffle.host.TruffleHostEnvironment;
+import jdk.graal.compiler.util.ObjectCopier;
+import jdk.internal.module.Modules;
+import jdk.vm.ci.hotspot.HotSpotJVMCIBackendFactory;
+import jdk.vm.ci.hotspot.HotSpotJVMCIRuntime;
+import jdk.vm.ci.services.JVMCIServiceLocator;
+import org.graalvm.collections.EconomicMap;
+import org.graalvm.jniutils.NativeBridgeSupport;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.FieldValueTransformer;
+import org.graalvm.nativeimage.hosted.RuntimeClassInitialization;
+import org.graalvm.nativeimage.hosted.RuntimeReflection;
+
+import jdk.graal.compiler.core.common.Fields;
+import jdk.graal.compiler.debug.GraalError;
+import jdk.graal.compiler.graph.Edges;
+import jdk.graal.compiler.options.OptionDescriptor;
+import jdk.graal.compiler.options.OptionKey;
+import jdk.graal.compiler.serviceprovider.GlobalAtomicLong;
+import jdk.graal.compiler.core.common.FeatureComponent;
+import jdk.graal.nativeimage.LibGraalLoader;
+import jdk.graal.nativeimage.hosted.GlobalData;
+import jdk.vm.ci.hotspot.HotSpotModifiers;
+import org.graalvm.nativeimage.hosted.RuntimeSystemProperties;
+
+/**
+ * This feature builds the libgraal shared library (e.g., libjvmcicompiler.so on linux).
+ */
+@Platforms(Platform.HOSTED_ONLY.class)
+public final class LibGraalFeature implements Feature {
+
+    /**
+     * Prefix to be used when {@linkplain RuntimeSystemProperties#register registering} properties
+     * describing the image configuration for libgraal. This is analogous to the configuration info
+     * displayed by {@code -XshowSettings}.
+     *
+     * For example:
+     *
+     * <pre>
+     * RuntimeSystemProperties.register(NATIVE_IMAGE_SETTING_KEY_PREFIX + "gc", "serial");
+     * </pre>
+     */
+    public static final String NATIVE_IMAGE_SETTING_KEY_PREFIX = "org.graalvm.nativeimage.setting.";
```

Comment:
`jdk.graal.compiler` exports everything (`*`) to `org.graalvm.nativeimage...` which makes of course sense for hosted graal usage. 

If I understand correctly, `LibGraalFeature` is guest graal only. I.e., `org.graalvm.nativeimage...` should not depend on it, or even have access to it, right? Otherwise, we might introduce dependencies from native image to a specific version of `LibGraalFeature` without noticing it.

Currently, we only access `NATIVE_IMAGE_SETTING_KEY_PREFIX` so not a problem. Still, would it make sense to separate it out `LibGraalFeature` to avoid problems?

Reply:
Moved to `jdk.graal.compiler.core.common.LibGraalSupport#NATIVE_IMAGE_SETTING_KEY_PREFIX`.